### ISSUE TITLE
Parse error at or near `RETURN_VALUE' instruction (Python 3.8)

### ISSUE THREAD

**Issue by abmyii** [rocky/python-decompile3#35]
Note, to reproduce this, the patch mentioned in https://github.com/rocky/python-uncompyle6/issues/310 has to be applied.

The parsing error given (inline - in the file attached below):
```
 L. 572       318  POP_TOP
              320  LOAD_STR                 '0'
              322  RETURN_VALUE
            324_0  COME_FROM           306  '306'

Parse error at or near `RETURN_VALUE' instruction at offset 322
```

Link to the output file: https://gofile.io/?c=36HCJy
Link to the `pyc` file: https://gofile.io/?c=MV8jCW

**Comment by abmyii** [comment]
```python
# decompyle3 version 3.3.2
# Python bytecode 3.8 (3413)
# Decompiled from: Python 3.8.1 (default, Dec 22 2019, 08:15:39) 
# [GCC 7.4.0]
# Embedded file name: Main.py
# Size of source mod 2**32: 21 bytes
Instruction context:
   
 L. 572       318  POP_TOP          
                 320  LOAD_STR                 '0'
->               322  RETURN_VALUE     
Traceback (most recent call last):
  File "//.local/bin/decompyle3", line 11, in <module>
    load_entry_point('decompyle3==3.3.2', 'console_scripts', 'decompyle3')()
  File "/.local/lib/python3.8/site-packages/decompyle3/bin/decompile.py", line 189, in main_bin
    result = main(
  File "/.local/lib/python3.8/site-packages/decompyle3/main.py", line 294, in main
    deparsed = decompile_file(
  File "/.local/lib/python3.8/site-packages/decompyle3/main.py", line 205, in decompile_file
    decompile(
  File "/.local/lib/python3.8/site-packages/decompyle3/main.py", line 137, in decompile
    deparsed = deparse_fn(
  File "/.local/lib/python3.8/site-packages/decompyle3/semantics/pysource.py", line 2250, in code_deparse
    deparsed.gen_source(deparsed.ast, co.co_name, customize)
  File "/.local/lib/python3.8/site-packages/decompyle3/semantics/pysource.py", line 2090, in gen_source
    self.text = self.traverse(ast, is_lambda=is_lambda)
  File "/.local/lib/python3.8/site-packages/decompyle3/semantics/pysource.py", line 420, in traverse
    self.preorder(node)
  File "/.local/lib/python3.8/site-packages/decompyle3/semantics/pysource.py", line 398, in preorder
    super(SourceWalker, self).preorder(node)
  File "/.local/lib/python3.8/site-packages/spark_parser/ast.py", line 117, in preorder
    self.preorder(kid)
  File "/.local/lib/python3.8/site-packages/decompyle3/semantics/pysource.py", line 398, in preorder
    super(SourceWalker, self).preorder(node)
  File "/.local/lib/python3.8/site-packages/spark_parser/ast.py", line 112, in preorder
    self.default(node)
  File "/.local/lib/python3.8/site-packages/decompyle3/semantics/pysource.py", line 1919, in default
    self.template_engine(table[key.kind], node)
  File "/.local/lib/python3.8/site-packages/decompyle3/semantics/pysource.py", line 1829, in template_engine
    self.preorder(node[index])
  File "/.local/lib/python3.8/site-packages/decompyle3/semantics/pysource.py", line 398, in preorder
    super(SourceWalker, self).preorder(node)
  File "/.local/lib/python3.8/site-packages/spark_parser/ast.py", line 117, in preorder
    self.preorder(kid)
  File "/.local/lib/python3.8/site-packages/decompyle3/semantics/pysource.py", line 398, in preorder
    super(SourceWalker, self).preorder(node)
  File "/.local/lib/python3.8/site-packages/spark_parser/ast.py", line 117, in preorder
    self.preorder(kid)
  File "/.local/lib/python3.8/site-packages/decompyle3/semantics/pysource.py", line 398, in preorder
    super(SourceWalker, self).preorder(node)
  File "/.local/lib/python3.8/site-packages/spark_parser/ast.py", line 117, in preorder
    self.preorder(kid)
  File "/.local/lib/python3.8/site-packages/decompyle3/semantics/pysource.py", line 398, in preorder
    super(SourceWalker, self).preorder(node)
  File "/.local/lib/python3.8/site-packages/spark_parser/ast.py", line 112, in preorder
    self.default(node)
  File "/.local/lib/python3.8/site-packages/decompyle3/semantics/pysource.py", line 1919, in default
    self.template_engine(table[key.kind], node)
  File "/.local/lib/python3.8/site-packages/decompyle3/semantics/pysource.py", line 1829, in template_engine
    self.preorder(node[index])
  File "/.local/lib/python3.8/site-packages/decompyle3/semantics/pysource.py", line 398, in preorder
    super(SourceWalker, self).preorder(node)
  File "/.local/lib/python3.8/site-packages/spark_parser/ast.py", line 117, in preorder
    self.preorder(kid)
  File "/.local/lib/python3.8/site-packages/decompyle3/semantics/pysource.py", line 398, in preorder
    super(SourceWalker, self).preorder(node)
  File "/.local/lib/python3.8/site-packages/spark_parser/ast.py", line 117, in preorder
    self.preorder(kid)
  File "/.local/lib/python3.8/site-packages/decompyle3/semantics/pysource.py", line 398, in preorder
    super(SourceWalker, self).preorder(node)
  File "/.local/lib/python3.8/site-packages/spark_parser/ast.py", line 117, in preorder
    self.preorder(kid)
  File "/.local/lib/python3.8/site-packages/decompyle3/semantics/pysource.py", line 398, in preorder
    super(SourceWalker, self).preorder(node)
  File "/.local/lib/python3.8/site-packages/spark_parser/ast.py", line 110, in preorder
    func(node)
  File "/.local/lib/python3.8/site-packages/decompyle3/semantics/customize37.py", line 831, in n_function_def
    self.default(node)
  File "/.local/lib/python3.8/site-packages/decompyle3/semantics/pysource.py", line 1919, in default
    self.template_engine(table[key.kind], node)
  File "/.local/lib/python3.8/site-packages/decompyle3/semantics/pysource.py", line 1829, in template_engine
    self.preorder(node[index])
  File "/.local/lib/python3.8/site-packages/decompyle3/semantics/pysource.py", line 398, in preorder
    super(SourceWalker, self).preorder(node)
  File "/.local/lib/python3.8/site-packages/spark_parser/ast.py", line 110, in preorder
    func(node)
  File "/.local/lib/python3.8/site-packages/decompyle3/semantics/pysource.py", line 818, in n_mkfunc
    make_function36(self, node, is_lambda=False, code_node=code_node)
  File "/.local/lib/python3.8/site-packages/decompyle3/semantics/make_function36.py", line 159, in make_function36
    ast = self.build_ast(
  File "/.local/lib/python3.8/site-packages/decompyle3/semantics/pysource.py", line 2155, in build_ast
    ast = python_parser.parse(self.p, tokens, customize, is_lambda=is_lambda)
  File "/.local/lib/python3.8/site-packages/decompyle3/parsers/main.py", line 351, in parse
    ast = p.parse(tokens)
  File "/.local/lib/python3.8/site-packages/spark_parser/spark.py", line 501, in parse
    self.error(tokens, i-1)
  File "/.local/lib/python3.8/site-packages/decompyle3/parsers/main.py", line 206, in error
    raise ParserError(err_token, err_token.offset, self.debug["reduce"])
decompyle3.parsers.main.ParserError: Parse error at or near `RETURN_VALUE' instruction at offset 322
```
Full error got by removing `try`/`except` functions, if it helps.

**Comment by rocky** [comment]
FIrst, again thanks for investigating. The process you use where you add custom debug instructions, such as `if offset == ...` is exactly what I do. (Instead of print though I use my trepan debugger, but the idea is the same).  LIkewise modifying the code to remove `try/exept` is a fine thing to do and I add or remove them with abandon. That's the great thing about a version-control system: you can make these wild changes and find and revert any or all of them easily. 

That said, this program is way to large to seriously try to fix directly. The code needs to be whittled down the smallest program that exhibits the problem. at 2.7K lines I am sure we can come up with something of less size than this that has the same problem. 

Just to set some groundwork and expectation, right now I've been more focused on 3.7 code than 3.8. Support for both of them is a bit weak, and 3.8 is weaker since it is based on differences from 3.7. And of course Python bytecode generation gets trickier over time. 

There is no shortage of bugs to fix. Just a shortage of people to help out or fix them. 

Therefore the only sane thing to do is to start out with easier well isolated bugs and them move on to the more difficult cases. If there are special difficult cases to fix, as mentioned above, I whittle down the program into the smallest thing program that has that bug. Of course I may whittle it down into a _different_ bug, but that's okay as far as I am concerned. Or there may be 5 different bugs but I narrowed it to one. Same thing.

With respect to the specific patch that you provided, while I appreciate the work you did to figure this out, what is needed is a more pervasive kind of fix. 

If you look at the scanner38.py a little before where you were looking, you'll see the comment: 

```
            # Turn JUMP opcodes into "BREAK_LOOP" opcodes.
            # FIXME: this should be replaced by proper control flow.
```

In the issue in `uncompyle6` I described that the underlying problem is that the code is working with this enhanced bytecode set where we've folded all `EXTENDED_ARGS` instructions in the instructiion that it is to be applied to. 

However even that, as I have come to learn, is not enough. 

So I've started raising the level of abstraction into a full-fledged instruction, not a bunch of bytecode bytes that form an instruction. This way we are isolated from opcode names changing on opcode values changing (all of which have occurred in the past). If we want to know if an instruction can fall through to the next instruction that can be marked a a separate field or property of the instruction.

To get a sense of how the particular code you looked at is addressed by working with instructions rather than the token-based instruction (which again is a little bit higher level than bytecode instructions), then see https://github.com/rocky/python-decompile3/pull/36

**Comment by abmyii** [comment]
Firstly, thank you very much for taking the time to write such detailed and kind comments!

I'll begin by addressing the points made in https://github.com/rocky/python-uncompyle6/issues/310#issuecomment-608170388:

> I just tried applying the change you suggested and while that no longer throws a KeyError exception, I am not getting a parse of the instructions and therefore no decompilation.

Yes, I neglecteed to mention that, though I fixed that problem in uncompyle6, a further issue occurred (`Parse error at or near `BEGIN_FINALLY' instruction at offset 366`). I realised it might be something to do with `uncompyle6` having poor support for 3.7+ so I applied the patch in an identical manner to `decompile3` and it worked, up to the error raised in this issue.

> I hope this answers the questions here. For what should be done, and moving towards addressing this let's move the discussion to decompiyle3 where I'll post the remainder.

Ah, I think I understand now, thanks to your great explanation!

> First, again thanks for investigating. The process you use where you add custom debug instructions, such as if offset == ... is exactly what I do. (Instead of print though I use my trepan debugger, but the idea is the same). LIkewise modifying the code to remove try/exept is a fine thing to do and I add or remove them with abandon. That's the great thing about a version-control system: you can make these wild changes and find and revert any or all of them easily.

No problem, I found it too intriguing a prospect to pass! I did find it a little difficult to locate where this error occurred due to the `try/except` statements "hiding" the details of where the error was occurring, etc. In the end I decide to remove all the catches for `ParseError` and was able to get the "full" error. I'm sure it would have helped to use a proper debugger and I will do so in the future now I know about it.

> That said, this program is way to large to seriously try to fix directly. The code needs to be whittled down the smallest program that exhibits the problem. at 2.7K lines I am sure we can come up with something of less size than this that has the same problem.

Yup, I've tried to, but I can't exactly reproduce it since I don't really understand what the disassembly is showing (i.e. I don't understand it enough to be able to reverse it manually). The code isn't my own and the author isn't sure - though they believe it returned a list, and also may have had some `%` formatting.

> Therefore the only sane thing to do is to start out with easier well isolated bugs and them move on to the more difficult cases. If there are special difficult cases to fix, as mentioned above, I whittle down the program into the smallest thing program that has that bug. Of course I may whittle it down into a different bug, but that's okay as far as I am concerned. Or there may be 5 different bugs but I narrowed it to one. Same thing.

Any tips on how to do so in this case?

> With respect to the specific patch that you provided, while I appreciate the work you did to figure this out, what is needed is a more pervasive kind of fix.

I totally understand - it was the quickest hack I found to move on from that error.

> That's the spirit! I applaud you. Alas after looking at this, it looks like it is not as simple as we would have liked...

Thank you!

> There is no shortage of bugs to fix. Just a shortage of people to help out or fix them.

> To get a sense of how the particular code you looked at is addressed by working with instructions rather than the token-based instruction (which again is a little bit higher level than bytecode instructions), then see #36

I am (very much) out of my depth as this is the first time I've looked into bytecode and disassembly, but I'll help where I can. Perhaps with some guidance from you I may be able to assist with some of the more intricate issues, if that isn't too inconvenient!

**Comment by rocky** [comment]
> Any tips on how to do so in this case?

If you have similar code that will help. Or a version that does decompile, or a version of the source code that is close, maybe an earlier or later version. Or the same code byte compiled to another Python version like 3.7. These will guide you. 

I've written a bytecode assembler which might help, but I don't want to go down that road here or offer guidance on that. It opens up a whole new can a worms. 

There is the decompiler service by crazycompilers mentioned in uncompyle6 that might be willing to take this on.

I could probably decompile this with this tool and by hand but it would be tortuous and something that would take a lot of time for me to do.  In a month or so I'll be offering sponsorship of this and the uncompyle6 project since this takes a lot of my time to do, and I no financial benefit from any of it. (Actually, it is a financial drain since it keeps me from doing other things where that I might get paid for)

> Perhaps with some guidance from you I may be able to assist with some of the more intricate issues, if that isn't too inconvenient!

In all of my open-source projects I welcome help at any or all levels. There are so many ways to improve this code, and so much to do that I basically, just leave off in refactoring and making small code improvements until I can't stand working around them and then decide it is time to address.

There is a _lot_ written about how these decompilers work. Espeically when you consider there is _nothing_ written about how its predecessors `unompyle2` and `decompyle` work nor any of the other Python decompilers. 

The best place to start is the [wiki for uncompyle6](https://github.com/rocky/python-uncompyle6/wiki), specificallly [How does this code work?](https://github.com/rocky/python-uncompyle6/wiki/How-does-this-code-work%3F)

**Comment by abmyii** [comment]
> If you have similar code that will help. Or a version that does decompile, or a version of the source code that is close, maybe an earlier or later version. Or the same code byte compiled to another Python version like 3.7. These will guide you.

I'll see what I can do.

> I've written a bytecode assembler which might help, but I don't want to go down that road here or offer guidance on that. It opens up a whole new can a worms.

> There is the decompiler service by crazycompilers mentioned in uncompyle6 that might be willing to take this on.

I believe they only support up to Python 2.7 (http://www.crazy-compilers.com/decompyle/service.html).

> I could probably decompile this with this tool and by hand but it would be tortuous and something that would take a lot of time for me to do. In a month or so I'll be offering sponsorship of this and the uncompyle6 project since this takes a lot of my time to do, and I no financial benefit from any of it. (Actually, it is a financial drain since it keeps me from doing other things where that I might get paid for)

Surely only that one statement needs to be reversed to understand the cause of the problem? I understand what you mean, though. I hope you'll get sponsorship for these amazing projects!

> The best place to start is the wiki for uncompyle6, specifically How does this code work?

I'll definitely take a look and see if I can wrap my head around it. If I am able to and I understand the code-base properly then perhaps I'll be able to contribute and improve the project.

**Comment by rocky** [comment]
> Surely only that one statement needs to be reversed to understand the cause of the problem?

No. If you hange just one bytecode value offset by 2 and the entire program structure can be changed. 

Of the things that causes a lot of problems are programs with EXTENDED_ARG and this program has a hundred or so of them. And then there is the problem that 3.8 is has the worst coverage right now;  there is no reason to believe there is only one bug in decompyle6 that will make this work?

**Comment by rocky** [comment]
Had some free time to look at old issuses today. However I see that the source code is not on the links above.  So I'll move onto something else.

**Comment by abmyii** [comment]
I can re-upload the files if necessary. I spent a week or so trying to debug it myself (using the excellent PyCharm debugger), but made no progress at all unfortunately, so I gave up on it. As you said, it's very difficult to find the offending line in such a massive file, so perhaps it's not worth the effort.

**Comment by rocky** [comment]
>  I can re-upload the files if necessary.

Nah - let's just close this. It's not worth the effort.

**Comment by abmyii** [comment]
Appreciate the help throughout - thanks again.

**Comment by abmyii** [comment]
Good news - the fix for #40 also fixes this! I still had to apply the "patch" mentioned in the original issue (https://github.com/rocky/python-uncompyle6/issues/310), but otherwise it worked fine. Thank you very much for your effort and help throughout!
